module/zfs/bpobj.c

```diff
@@ -206,155 +206,213 @@ bpobj_is_empty(bpobj_t *bpo)
-			    FTAG, &dbuf, 0);
+			blkptr_t *bparray = dbuf->db_data;
+			blkptr_t *bp = &bparray[blkoff];
+			err = func(arg, bp, tx);
 			if (err)
 				break;
+			if (free) {
+				spa_t *spa = dmu_objset_spa(bpo->bpo_os);
+				freed += bp_get_dsize_sync(spa, bp);
+				bpo->bpo_phys->bpo_bytes -=
+				    bp_get_dsize_sync(spa, bp);
+				ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
+				if (bpo->bpo_havecomp) {
+					bpo->bpo_phys->bpo_comp -=
+					    BP_GET_PSIZE(bp);
+					bpo->bpo_phys->bpo_uncomp -=
+					    BP_GET_UCSIZE(bp);
+					comp_freed += BP_GET_PSIZE(bp);
+					uncomp_freed += BP_GET_UCSIZE(bp);
+				}
+				bpo->bpo_phys->bpo_num_blkptrs--;
+				ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs,
+				    >=, 0);
+			}
+		}
+		if (dbuf) {
+			dmu_buf_rele(dbuf, FTAG);
+			dbuf = NULL;
 		}
-
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		bparray = dbuf->db_data;
-		bp = &bparray[blkoff];
-		err = func(arg, bp, tx);
-		if (err)
-			break;
 		if (free) {
-			bpo->bpo_phys->bpo_bytes -=
-			    bp_get_dsize_sync(dmu_objset_spa(bpo->bpo_os), bp);
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			if (bpo->bpo_havecomp) {
-				bpo->bpo_phys->bpo_comp -= BP_GET_PSIZE(bp);
-				bpo->bpo_phys->bpo_uncomp -= BP_GET_UCSIZE(bp);
+			VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
+			    bpo->bpo_object, (i + 1) * sizeof (blkptr_t),
+			    DMU_OBJECT_END, tx));
+		}
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				if (free) {
+					bpobj_phys_t *bpop =
+					    bpi->bpi_parent->bpo_phys;
+					bpop->bpo_bytes -= freed;
+					ASSERT3S(bpop->bpo_bytes, >=, 0);
+					bpop->bpo_comp -= comp_freed;
+					bpop->bpo_uncomp -= uncomp_freed;
+				}
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
 			}
-			bpo->bpo_phys->bpo_num_blkptrs--;
-			ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs, >=, 0);
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			goto out;
 		}
-	}
-	if (dbuf) {
-		dmu_buf_rele(dbuf, FTAG);
-		dbuf = NULL;
-	}
-	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
-	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
-
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
-	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
 
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
+		if (!bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0) {
+			/* propogate reduction in space used and clean up */
+			if (free && (bpi->bpi_parent != NULL)) {
```

Comment:
Do we need to propagate this to our older ancestors (grand-parent, great-grand-parent, etc)?

Reply:
Here's some debug output.  bpobj ffff8ef8f98170d0 is the parent of all 34 subobjs.  I'd like to prove that the flattened list would always propagate the accounting even if the test generated multiple levels.  Experimented with clones but that didn't help, all subobjs were still children of dp_free_bpobj.
```
[  425.942985] NOTICE: bpi bpo ffff8ef8f98170d0 parent           (null) index 0 bytes 158490624 (before)
[  425.943103] NOTICE: bpi bpo ffff8ef8fa18d420 parent ffff8ef8f98170d0 index 34 bytes 4528128 (before)
[  425.977424] NOTICE: bpi bpo ffff8ef8fa18d420 parent ffff8ef8f98170d0 index 34 bytes 0 parent bytes 153962496 (after)
[  425.977428] NOTICE: bpi bpo ffff8ef8fa18d840 parent ffff8ef8f98170d0 index 33 bytes 4528128 (before)
[  426.010388] NOTICE: bpi bpo ffff8ef8fa18d840 parent ffff8ef8f98170d0 index 33 bytes 0 parent bytes 149434368 (after)
[  426.010391] NOTICE: bpi bpo ffff8ef8fa18dba0 parent ffff8ef8f98170d0 index 32 bytes 4528128 (before)
[  426.042430] NOTICE: bpi bpo ffff8ef8fa18dba0 parent ffff8ef8f98170d0 index 32 bytes 0 parent bytes 144906240 (after)
[  426.042434] NOTICE: bpi bpo ffff8ef8d2c04ba0 parent ffff8ef8f98170d0 index 31 bytes 4528128 (before)
[  426.076430] NOTICE: bpi bpo ffff8ef8d2c04ba0 parent ffff8ef8f98170d0 index 31 bytes 0 parent bytes 140378112 (after)
[  426.076433] NOTICE: bpi bpo ffff8ef8d2c04e40 parent ffff8ef8f98170d0 index 30 bytes 4528128 (before)
[  426.110688] NOTICE: bpi bpo ffff8ef8d2c04e40 parent ffff8ef8f98170d0 index 30 bytes 0 parent bytes 135849984 (after)
[  426.110690] NOTICE: bpi bpo ffff8ef8d2c04660 parent ffff8ef8f98170d0 index 29 bytes 4528128 (before)
[  426.139851] NOTICE: bpi bpo ffff8ef8d2c04660 parent ffff8ef8f98170d0 index 29 bytes 0 parent bytes 131321856 (after)
[  426.139855] NOTICE: bpi bpo ffff8ef8d2c041e0 parent ffff8ef8f98170d0 index 28 bytes 4528128 (before)
[  426.181628] NOTICE: bpi bpo ffff8ef8d2c041e0 parent ffff8ef8f98170d0 index 28 bytes 0 parent bytes 126793728 (after)
[  426.181631] NOTICE: bpi bpo ffff8ef8d2c04000 parent ffff8ef8f98170d0 index 27 bytes 4528128 (before)
[  426.214137] NOTICE: bpi bpo ffff8ef8d2c04000 parent ffff8ef8f98170d0 index 27 bytes 0 parent bytes 122265600 (after)
[  426.214140] NOTICE: bpi bpo ffff8ef8d2c04780 parent ffff8ef8f98170d0 index 26 bytes 4528128 (before)
[  426.247009] NOTICE: bpi bpo ffff8ef8d2c04780 parent ffff8ef8f98170d0 index 26 bytes 0 parent bytes 117737472 (after)
[  426.247012] NOTICE: bpi bpo ffff8ef8d2c04900 parent ffff8ef8f98170d0 index 25 bytes 4528128 (before)
[  426.293723] NOTICE: bpi bpo ffff8ef8d2c04900 parent ffff8ef8f98170d0 index 25 bytes 0 parent bytes 113209344 (after)
[  426.293726] NOTICE: bpi bpo ffff8ef8d2c04a80 parent ffff8ef8f98170d0 index 24 bytes 4528128 (before)
[  426.361108] NOTICE: bpi bpo ffff8ef8d2c04a80 parent ffff8ef8f98170d0 index 24 bytes 0 parent bytes 108681216 (after)
[  426.361111] NOTICE: bpi bpo ffff8ef8d2c04ea0 parent ffff8ef8f98170d0 index 23 bytes 4528128 (before)
[  426.426093] NOTICE: bpi bpo ffff8ef8d2c04ea0 parent ffff8ef8f98170d0 index 23 bytes 0 parent bytes 104153088 (after)
[  426.426096] NOTICE: bpi bpo ffff8ef8d2c04d20 parent ffff8ef8f98170d0 index 22 bytes 4528128 (before)
[  426.495193] NOTICE: bpi bpo ffff8ef8d2c04d20 parent ffff8ef8f98170d0 index 22 bytes 0 parent bytes 99624960 (after)
[  426.495196] NOTICE: bpi bpo ffff8ef8d2c04a20 parent ffff8ef8f98170d0 index 21 bytes 4528128 (before)
[  426.530717] NOTICE: bpi bpo ffff8ef8d2c04a20 parent ffff8ef8f98170d0 index 21 bytes 0 parent bytes 95096832 (after)
[  426.530720] NOTICE: bpi bpo ffff8ef8d2c04540 parent ffff8ef8f98170d0 index 20 bytes 4528128 (before)
[  426.593854] NOTICE: bpi bpo ffff8ef8d2c04540 parent ffff8ef8f98170d0 index 20 bytes 0 parent bytes 90568704 (after)
[  426.593857] NOTICE: bpi bpo ffff8ef8d2c04420 parent ffff8ef8f98170d0 index 19 bytes 4530176 (before)
[  426.648729] NOTICE: bpi bpo ffff8ef8d2c04420 parent ffff8ef8f98170d0 index 19 bytes 0 parent bytes 86038528 (after)
[  426.648732] NOTICE: bpi bpo ffff8ef8d2c04240 parent ffff8ef8f98170d0 index 18 bytes 4528128 (before)
[  426.691102] NOTICE: bpi bpo ffff8ef8d2c04240 parent ffff8ef8f98170d0 index 18 bytes 0 parent bytes 81510400 (after)
[  426.691105] NOTICE: bpi bpo ffff8ef8d2c04b40 parent ffff8ef8f98170d0 index 17 bytes 4528128 (before)
[  426.749585] NOTICE: bpi bpo ffff8ef8d2c04b40 parent ffff8ef8f98170d0 index 17 bytes 0 parent bytes 76982272 (after)
[  426.749589] NOTICE: bpi bpo ffff8ef8d2c04c60 parent ffff8ef8f98170d0 index 16 bytes 4528128 (before)
[  426.791471] NOTICE: bpi bpo ffff8ef8d2c04c60 parent ffff8ef8f98170d0 index 16 bytes 0 parent bytes 72454144 (after)
[  426.791474] NOTICE: bpi bpo ffff8ef8d2c040c0 parent ffff8ef8f98170d0 index 15 bytes 4528128 (before)
[  426.851609] NOTICE: bpi bpo ffff8ef8d2c040c0 parent ffff8ef8f98170d0 index 15 bytes 0 parent bytes 67926016 (after)
[  426.851613] NOTICE: bpi bpo ffff8ef8d2c04120 parent ffff8ef8f98170d0 index 14 bytes 4528128 (before)
[  426.901299] NOTICE: bpi bpo ffff8ef8d2c04120 parent ffff8ef8f98170d0 index 14 bytes 0 parent bytes 63397888 (after)
[  426.901304] NOTICE: bpi bpo ffff8ef8d2c04480 parent ffff8ef8f98170d0 index 13 bytes 4528128 (before)
[  426.934305] NOTICE: bpi bpo ffff8ef8d2c04480 parent ffff8ef8f98170d0 index 13 bytes 0 parent bytes 58869760 (after)
[  426.934308] NOTICE: bpi bpo ffff8ef8d2c044e0 parent ffff8ef8f98170d0 index 12 bytes 4528128 (before)
[  426.948021] NOTICE: error: 85
[  426.948032] NOTICE: returning with err 85 initial bpo bytes 56972288
[  426.948233] NOTICE: bpi bpo ffff8ef959c8e588 parent           (null) index 0 bytes 178176 (before)
[  426.948360] NOTICE: bpi bpo ffff8ef959c8e588 parent           (null) index 0 bytes 0 parent bytes 0 (after)
[  426.948361] NOTICE: returning with err 0 initial bpo bytes 0
[  427.514926] NOTICE: bpi bpo ffff8ef8f98170d0 parent           (null) index 0 bytes 56972288 (before)
[  427.514970] NOTICE: bpi bpo ffff8ef8e469ed80 parent ffff8ef8f98170d0 index 12 bytes 2630656 (before)
[  427.534730] NOTICE: bpi bpo ffff8ef8e469ed80 parent ffff8ef8f98170d0 index 12 bytes 0 parent bytes 54341632 (after)
[  427.534736] NOTICE: bpi bpo ffff8ef8e469e5a0 parent ffff8ef8f98170d0 index 11 bytes 4528128 (before)
[  427.577647] NOTICE: bpi bpo ffff8ef8e469e5a0 parent ffff8ef8f98170d0 index 11 bytes 0 parent bytes 49813504 (after)
[  427.577651] NOTICE: bpi bpo ffff8ef8e469e3c0 parent ffff8ef8f98170d0 index 10 bytes 4528128 (before)
[  427.613637] NOTICE: bpi bpo ffff8ef8e469e3c0 parent ffff8ef8f98170d0 index 10 bytes 0 parent bytes 45285376 (after)
[  427.613640] NOTICE: bpi bpo ffff8ef8e469eb40 parent ffff8ef8f98170d0 index 9 bytes 4530176 (before)
[  427.657938] NOTICE: bpi bpo ffff8ef8e469eb40 parent ffff8ef8f98170d0 index 9 bytes 0 parent bytes 40755200 (after)
[  427.657942] NOTICE: bpi bpo ffff8ef8e469e480 parent ffff8ef8f98170d0 index 8 bytes 4528128 (before)
[  427.700679] NOTICE: bpi bpo ffff8ef8e469e480 parent ffff8ef8f98170d0 index 8 bytes 0 parent bytes 36227072 (after)
[  427.700682] NOTICE: bpi bpo ffff8ef8e469e060 parent ffff8ef8f98170d0 index 7 bytes 4528128 (before)
[  427.738197] NOTICE: bpi bpo ffff8ef8e469e060 parent ffff8ef8f98170d0 index 7 bytes 0 parent bytes 31698944 (after)
[  427.738201] NOTICE: bpi bpo ffff8ef8e469e600 parent ffff8ef8f98170d0 index 6 bytes 4528128 (before)
[  427.779463] NOTICE: bpi bpo ffff8ef8e469e600 parent ffff8ef8f98170d0 index 6 bytes 0 parent bytes 27170816 (after)
[  427.779466] NOTICE: bpi bpo ffff8ef8e469e240 parent ffff8ef8f98170d0 index 5 bytes 4528128 (before)
[  427.821228] NOTICE: bpi bpo ffff8ef8e469e240 parent ffff8ef8f98170d0 index 5 bytes 0 parent bytes 22642688 (after)
[  427.821232] NOTICE: bpi bpo ffff8ef8e469e4e0 parent ffff8ef8f98170d0 index 4 bytes 4528128 (before)
[  427.858871] NOTICE: bpi bpo ffff8ef8e469e4e0 parent ffff8ef8f98170d0 index 4 bytes 0 parent bytes 18114560 (after)
[  427.858876] NOTICE: bpi bpo ffff8ef8e469e360 parent ffff8ef8f98170d0 index 3 bytes 4529152 (before)
[  427.897173] NOTICE: bpi bpo ffff8ef8e469e360 parent ffff8ef8f98170d0 index 3 bytes 0 parent bytes 13585408 (after)
[  427.897176] NOTICE: bpi bpo ffff8ef8e469e000 parent ffff8ef8f98170d0 index 2 bytes 4528128 (before)
[  427.943388] NOTICE: bpi bpo ffff8ef8e469e000 parent ffff8ef8f98170d0 index 2 bytes 0 parent bytes 9057280 (after)
[  427.943426] NOTICE: bpi bpo ffff8ef8e469ecc0 parent ffff8ef8f98170d0 index 1 bytes 4529152 (before)
[  427.981859] NOTICE: bpi bpo ffff8ef8e469ecc0 parent ffff8ef8f98170d0 index 1 bytes 0 parent bytes 4528128 (after)
[  427.981862] NOTICE: bpi bpo ffff8ef8e469eba0 parent ffff8ef8f98170d0 index 0 bytes 4528128 (before)
[  428.017620] NOTICE: bpi bpo ffff8ef8e469eba0 parent ffff8ef8f98170d0 index 0 bytes 0 parent bytes 0 (after)
[  428.017623] NOTICE: returning with err 0 initial bpo bytes 0
```